WarpX: Remove Deprecated Versions

ax3l opened the pull request:
This is necessary because we improved and updated the build system a lot over the last 4 years and I do not want to maintain quirks for older versions anymore to stay productive. For users, we anyway only support the latest releases.

## To Do

- [x] rebase after #46763 was merged
- [x] simplify `pip_install_nodeps` via the new 24.09+ options in [warpx](https://github.com/ECP-WarpX/WarpX/pull/5128) and [py-amrex](https://github.com/AMReX-Codes/pyamrex/pull/353)
- [x] fix #49546

Can I cleverly use our new CTest support for testing?
<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

ax3l commented:
@spackbot fix style

EZoni commented on var/spack/repos/builtin/packages/warpx/package.py:
```diff
@@ -332,57 +165,13 @@ class Warpx(CMakePackage, PythonExtension):
         depends_on("py-periodictable@1.5:1", type=("build", "run"))
-        depends_on("py-picmistandard@0.28.0", type=("build", "run"), when="@23.11:24.07")
-        depends_on("py-picmistandard@0.29.0", type=("build", "run"), when="@24.08")
         depends_on("py-picmistandard@0.30.0", type=("build", "run"), when="@24.09:")
```
Should this be
```suggestion
        depends_on("py-picmistandard@0.30.0", type=("build", "run"))
```
since the minimum version is now `24.10` anyways (if I'm not mistaken)?

ax3l replied:
You are correct.

I kept this so we do not forget to update and constrain it once we add the next release. We will have to change this line then anyway.

Depending on when we change PICMI the next time, this reads something like
```suggestion
        depends_on("py-picmistandard@0.30.0", type=("build", "run"), when="@24.09:24.11")
        depends_on("py-picmistandard@0.31.0", type=("build", "run"), when="@24.12:")
```
(note: Spack uses _inclusive_ upper version bounds)

eugeneswalker reviewed (dismissed):
Thanks!

bernhardkaindl commented on var/spack/repos/builtin/packages/warpx/package.py:
```diff
@@ -423,43 +209,35 @@ def cmake_args(self):
+            pip_args = PythonPipBuilder.std_args(self)
+            pip_args_install_index = pip_args.index("install")
+            pip_args_before_install = pip_args[:pip_args_install_index]
+            pip_args_after_install = pip_args[pip_args_install_index + 1 :] + [
+                f"--prefix={self.prefix}"
+            ]
+
+            args.append("-DWarpX_pyamrex_internal=OFF")
+            args.append("-DWarpX_pybind11_internal=OFF")
+            args.append(self.define_from_variant("WarpX_PYTHON_IPO", "python_ipo"))
+            # Additional parameters to pass to `pip`
+            args.append('-DPY_PIP_OPTIONS="' + " ".join(pip_args_before_install) + '"'),
+            # Additional parameters to pass to `pip install`
+            args.append('-DPY_PIP_INSTALL_OPTIONS="' + " ".join(pip_args_after_install) + '"'),
```
This causes:
```py
python3 -m pip "-vvv\ --no-input\ --no-cache-dir\ --disable-pip-version-check" wheel --no-build-isolation --no-deps --wheel-dir=amrex-whl ...
no such option: -\
```
This is caused by `-vvv\`. Those `\` are added by the cmake build to escape the spaces.

This means that you must remove the quotes `"` around the string that you add:
Those prevent these backslashes from getting consumed by escape expansion, and thus they reach `pip`, which you do not want.

In GitLab CI, I also saw in https://gitlab.spack.io/spack/spack/-/jobs/12901412:
```py
Error: Failed to install py-amrex due to TypeError: write() argument must be str, not list
```
To make this code more readable, consistently use f-strings instead of string concatenation.

bernhardkaindl reviewed (changes requested):
Hi, I've marked this PR as Draft now to prevent simple pushes to it to trigger the GitLab CI runners to run this PR. I think it would be great if it could be kept so until all packages in the PR are building in manual builds using `spack install <spec>` or in an environment and only be marked as ready for review if initial reviews also consider this PR as ready for the final testing in GitLab CI pipelines.

GitLab CI pipelines are currently lagging behind the develop branch and many PRs that are clean to build and approved cannot be merged solely because GitLab PR is lagging behind the develop branch. I therefore would like to ask to (voluntarily) conserving the use GitLab CI runner resources for PRs that are ready for the final verification for the merge and should have a very high chance of not causing GitLab CI pipelines to fail.

bernhardkaindl commented on var/spack/repos/builtin/packages/warpx/package.py:
```diff
@@ -423,43 +209,36 @@ def cmake_args(self):
-                args.append("-DWarpX_pyamrex_internal=OFF")
-                args.append("-DWarpX_pybind11_internal=OFF")
-                args.append(self.define_from_variant("WarpX_PYTHON_IPO", "python_ipo"))
+            pip_args = PythonPipBuilder.std_args(self) + [f"--prefix={self.prefix}"]
```
I pushed the equivalent already for `py-amrex`:

Fix errors passing `pip` arguments using `cmake` to pip:
- In CMake, argument lists are semicolon-delimited.
- Contrary to a shell, CMake preserves spaces and even escapes them automatically using backslashes.
  - When passing spaces between options, this `cmake` passes `pip` a long option that has spaces in it!
- That means when you set a CMake variable that contains spaces, they are passed as a stringle, long string. They are not split by spaces!
  - CMake are no does not do splitting by spaces, but by semilcolon instead.
- Thus, the only way to pass a list of multiple argument to a called command like `pip` is to pass them to CMake as a CMake list that is semicolon-separated:

After the 1st commit by me, this is the 2nd fix. If there are no other issues like that, this may fix GitLab.

I'll approve when this applied and a manual build passes.
- I would like to suggest leaving the status as Draft until the reviewers signalled that they would approve the removed version and the related changes. And I suggest setting it to ready when the PR is ready for a the final GitLab CI pipeline to create the condition for the merge.

ax3l replied:
Thanks. I arrived at the same conclusion yesterday.

This needs a mainline fix and patch first, because I do not treat them as lists yet. Will push more once this works.

ax3l commented:
> Hi, I've marked this PR as Draft now to prevent simple pushes to it to trigger the GitLab CI runners to run this PR. I think it would be great if it could be kept so until all packages in the PR are building in manual builds using `spack install <spec>` or in an environment and only be marked as ready for review if initial reviews also consider this PR as ready for the final testing in GitLab CI pipelines.
> 
> GitLab CI pipelines are currently lagging behind the develop branch and many PRs that are clean to build and approved cannot be merged solely because GitLab PR is lagging behind the develop branch. I therefore would like to ask to (voluntarily) conserving the use GitLab CI runner resources for PRs that are ready for the final verification for the merge and should have a very high chance of not causing GitLab CI pipelines to fail.

Sorry again as replied on Slack: I was in the impression the things I pushed worked, but was tricked by my local cache building an earlier version. I also thought a quick push after a non-passed build would cancel the queued build and only the latest HEAD of the PR gets build?

ax3l commented:
We now use oneAPI (oneMKL) FFTs for SYCL, let's see why this tries to pull in FFTW3 (that we do not need for SYCL) for E4S...

ax3l commented:
@WeiqunZhang can we make AMReX with SYCL compile with EB on even if we build 1D? That would simplify the multi-D builds for E4S deployment on SYCL.

Currently in CI,  I see:
```
/tmp/root/spack-stage/spack-stage-warpx-24.10-kixetezzwfx7xcusi6kmlio4kgnlugdy/spack-src/Source/EmbeddedBoundary/DistanceToEB.H:124:5: error: SYCL kernel cannot call an undefined function without SYCL_EXTERNAL attribute
  124 |     WARPX_ABORT_WITH_MESSAGE("Error: interp_distance not yet implemented in 1D");
      |     ^
/tmp/root/spack-stage/spack-stage-warpx-24.10-kixetezzwfx7xcusi6kmlio4kgnlugdy/spack-src/Source/Utils/TextMsg.H:15:64: note: expanded from macro 'WARPX_ABORT_WITH_MESSAGE'
   15 | #define WARPX_ABORT_WITH_MESSAGE(MSG) ablastr::utils::TextMsg::Abort( __FILE__, __LINE__ , MSG)
      |                                                                ^
/tmp/root/spack-stage/spack-stage-warpx-24.10-kixetezzwfx7xcusi6kmlio4kgnlugdy/spack-src/Source/ablastr/utils/TextMsg.H:71:5: note: 'Abort' declared here
   71 |     Abort (const char *file, int line, const std::string &msg);
      |     ^
/tmp/root/spack-stage/spack-stage-warpx-24.10-kixetezzwfx7xcusi6kmlio4kgnlugdy/spack-src/Source/EmbeddedBoundary/DistanceToEB.H:43:17: note: called by 'interp_normal'
   43 | amrex::RealVect interp_normal (int i, int j, int k, const amrex::Real W[AMREX_SPACEDIM][2],
      |                 ^
In file included from /tmp/root/spack-stage/spack-stage-warpx-24.10-kixetezzwfx7xcusi6kmlio4kgnlugdy/spack-src/Source/Particles/MultiParticleContainer.cpp:12:
In file included from /tmp/root/spack-stage/spack-stage-warpx-24.10-kixetezzwfx7xcusi6kmlio4kgnlugdy/spack-src/Source/Particles/MultiParticleContainer.H:18:
In file included from /tmp/root/spack-stage/spack-stage-warpx-24.10-kixetezzwfx7xcusi6kmlio4kgnlugdy/spack-src/Source/Particles/Collision/CollisionHandler.H:10:
In file included from /tmp/root/spack-stage/spack-stage-warpx-24.10-kixetezzwfx7xcusi6kmlio4kgnlugdy/spack-src/Source/Particles/Collision/CollisionBase.H:13:
In file included from /home/software/spack/[padded-to-256-chars]/linux-ubuntu22.04-x86_64_v3/oneapi-2024.2.1/amrex-24.10-a3ah5qshaxv67nswtcx7kwofq6atgfur/include/AMReX_Vector.H:5:
In file included from /home/software/spack/[padded-to-256-chars]/linux-ubuntu22.04-x86_64_v3/oneapi-2024.2.1/amrex-24.10-a3ah5qshaxv67nswtcx7kwofq6atgfur/include/AMReX_BLassert.H:6:
In file included from /home/software/spack/[padded-to-256-chars]/linux-ubuntu22.04-x86_64_v3/oneapi-2024.2.1/amrex-24.10-a3ah5qshaxv67nswtcx7kwofq6atgfur/include/AMReX.H:6:
In file included from /home/software/spack/[padded-to-256-chars]/linux-ubuntu22.04-x86_64_v3/oneapi-2024.2.1/amrex-24.10-a3ah5qshaxv67nswtcx7kwofq6atgfur/include/AMReX_GpuQualifiers.H:64:
In file included from /opt/intel/oneapi/compiler/2024.2/bin/compiler/../../include/sycl/sycl.hpp:25:
In file included from /opt/intel/oneapi/compiler/2024.2/bin/compiler/../../include/sycl/accessor.hpp:11:
In file included from /opt/intel/oneapi/compiler/2024.2/bin/compiler/../../include/sycl/access/access.hpp:14:
In file included from /opt/intel/oneapi/compiler/2024.2/bin/compiler/../../include/sycl/CL/__spirv/spirv_ops.hpp:25:
In file included from /opt/intel/oneapi/compiler/2024.2/bin/compiler/../../include/sycl/CL/__spirv/spirv_types.hpp:26:
In file included from /opt/intel/oneapi/compiler/2024.2/bin/compiler/../../include/sycl/half_type.hpp:13:
In file included from /opt/intel/oneapi/compiler/2024.2/bin/compiler/../../include/sycl/detail/iostream_proxy.hpp:11:
In file included from /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/istream:38:
In file included from /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/ios:42:
In file included from /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/ios_base.h:41:
In file included from /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/locale_classes.h:40:
In file included from /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/string:56:
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/basic_string.tcc:224:2: error: SYCL kernel cannot use exceptions
  224 |         __try
      |         ^
```

ax3l commented:
Oh, that looks like a WarpX 1D thing to fix.

alecbcs commented:
Merging in develop to trigger a new CI run based on updated binaries from develop.

bernhardkaindl reviewed (commented):
@ax3l: @24.10 fails to build in CI due to FFTW symbol errors:
https://gitlab.spack.io/spack/spack/-/jobs/13122361

ax3l commented on var/spack/repos/builtin/packages/warpx/package.py:
```diff
@@ -24,161 +24,31 @@ class Warpx(CMakePackage, PythonExtension):
-        deprecated=True,
+
+    # CMake: No FFTW Needed for SYCL anymore
+    patch(
+        "https://github.com/ECP-WarpX/WarpX/pull/5380.patch?full_index=1",
```
For this to work we will also need to patch in https://github.com/ECP-WarpX/WarpX/pull/5127
```suggestion
    patch(
        "https://github.com/ECP-WarpX/WarpX/pull/5127.patch?full_index=1",
        sha256="d2e112371582435399682bf7c1498e6ec04a36d62b2dbce329d566b862074c16",
        when="@24.10",
    )
    patch(
        "https://github.com/ECP-WarpX/WarpX/pull/5380.patch?full_index=1",
```

ax3l committed:
WarpX: Remove Deprecated Versions

alalazo commented:
There seems to be issues with the OneAPI build: https://gitlab.spack.io/spack/spack/-/jobs/15678403

ax3l committed:
Conflict: WarpX SYCL RZ FFT Issue

Conflict out on WarpX issue until fixed
https://github.com/BLAST-WarpX/warpx/issues/5774

ax3l committed:
Fix #49546

eugeneswalker commented:
This fixes https://github.com/spack/spack/issues/49546! Thanks!

```
==> Installing warpx-25.03-ypzorfclbmuz2ixvnyz2rnt5xvpho6zl [45/45]
==> No binary for warpx-25.03-ypzorfclbmuz2ixvnyz2rnt5xvpho6zl found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/ec/ec0a032fd2c838acec4b251fa3a8180b47186cc2216bc1adc6242cd275977590.tar.gz
==> No patches needed for warpx
==> warpx: Executing phase: 'cmake'
==> warpx: Executing phase: 'build'
==> warpx: Executing phase: 'install'
==> warpx: Successfully installed warpx-25.03-ypzorfclbmuz2ixvnyz2rnt5xvpho6zl
  Stage: 0.23s.  Cmake: 5.31s.  Build: 7m 6.02s.  Install: 2.71s.  Post-install: 0.13s.  Total: 7m 14.67s
[+] /spack/opt/spack/[padded-to-256-chars]/linux-ubuntu22.04-x86_64_v3/gcc-11.4.0/warpx-25.03-ypzorfclbmuz2ixvnyz2rnt5xvpho6zl
```
